packages/compartment-mapper/src/policy.js

```diff
@@ -0,0 +1,305 @@
+};
+
+/**
+ * Const string to identify the internal attenuators compartment
+ */
+export const ATTENUATORS_COMPARTMENT = '<ATTENUATORS>';
+
+const attenuatorsCache = new WeakMap();
+/**
+ * Goes through policy and lists all attenuator specifiers used.
+ * Memoization keyed on policy object reference
+ *
+ * @param {*} policy
+ * @returns {Array<string>} attenuators
+ */
+export const detectAttenuators = policy => {
+  if (!attenuatorsCache.has(policy)) {
+    const attenuators = [];
+    // a free recursive visitor implementation for simple objects
+    stringify(policy, (key, value) => {
+      if (key === 'attenuate' && typeof value === 'string') {
+        attenuators.push(value);
+      }
+      return value;
+    });
+    attenuatorsCache.set(policy, attenuators);
+  }
+  return attenuatorsCache.get(policy);
+};
+
+/**
+ * Generates a string identifying a package for policy lookup purposes.
+ *
+ * @param {PolicyIdentityComponents} identityComponents
+ * @returns {string}
+ */
+const generateCanonicalName = ({ isEntry = false, name, path }) => {
+  if (isEntry) {
+    return '<root>';
+  }
+  if (name === ATTENUATORS_COMPARTMENT) {
+    return ATTENUATORS_COMPARTMENT;
+  }
+  return path.join('>');
+};
+
+/**
+ * Verifies if a module identified by identityComponents can be a dependency of a package per packagePolicy
+ * packagePolicy is required, when policy is not set, skipping needs to be handled by the caller
+ *
+ * @param {PolicyIdentityComponents} identityComponents
+ * @param {*} packagePolicy
+ * @returns {boolean}
+ */
+export const dependencyAllowedByPolicy = (
+  identityComponents,
+  packagePolicy,
+) => {
+  const canonicalName = generateCanonicalName(identityComponents);
+  return packagePolicy.packages && !!packagePolicy.packages[canonicalName];
+};
+
+const validateDependencies = (policy, canonicalName) => {
+  const packages = policy.resources[canonicalName].packages;
+  if (!packages) {
+    return;
+  }
+
+  const packageIds = keys(packages);
+  const attenuators = detectAttenuators(policy);
+  // join attenuators with packageIds into a Set to deduplicate and check if all are listed in policy.resources
+  const allSpecifiers = new Set([...packageIds, ...attenuators]);
+  for (const specifier of allSpecifiers) {
+    if (!(specifier in policy.resources)) {
+      throw Error(
+        `Package "${specifier}" is allowed for "${canonicalName}" to import but its policy is not defined. Please add a policy for "${specifier}"`,
+      );
+    }
+  }
+};
+
+/**
+ * Returns the policy applicable to the canonicalName - either by taking from user
+ * supplied policy or returning localPolicy if user didn't specify one at runtime.
+ *
+ * @param {PolicyIdentityComponents} identityComponents - a key in the policy resources spec
+ * @param {Object|undefined} policy - user supplied policy
+ * @returns {Object|undefined} policy fragment if policy was specified
+ */
+export const getPolicyFor = (identityComponents, policy) => {
```

Comment:
Consider something like:
```suggestion
export const policyForCanonicalName = (identityComponents, policy) => {
```

Reply:
canonicalName is internal to policy.js - I'll try to come up with something.

Reply:
renamed to getPolicyForPackage